pkg/steps/artifacts.go

```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
 func (nopNotifier) Done(_ string) bool              { return true }
 func (nopNotifier) Cancel()                         {}
 
+// TestCaseNotifier allows a caller to generate per container unit tests
+// that provide better granularity for debugging problems. It intercepts
+// notifications and remembers the last pod retrieved.
+type TestCaseNotifier struct {
+	nested  ContainerNotifier
+	lastPod *coreapi.Pod
+}
+
+func NewTestCaseNotifier(nested ContainerNotifier) *TestCaseNotifier {
+	return &TestCaseNotifier{nested: nested}
+}
+
+func (n *TestCaseNotifier) Notify(pod *coreapi.Pod, containerName string) {
+	n.nested.Notify(pod, containerName)
+	n.lastPod = pod
+}
+
+func (n *TestCaseNotifier) Complete(podName string)  { n.nested.Complete(podName) }
+func (n *TestCaseNotifier) Done(podName string) bool { return n.nested.Done(podName) }
+func (n *TestCaseNotifier) Cancel()                  { n.nested.Cancel() }
+
+// SubTests returns one junit test for each terminated container with a name
+// in the annotation 'ci-operator.openshift.io/container-sub-tests' in the pod.
+// Invoking SubTests clears the last pod, so subsequent calls will return no
+// tests unless Notify() has been called in the meantime.
+func (n *TestCaseNotifier) SubTests(prefix string) []*junit.TestCase {
+	if n.lastPod == nil {
+		return nil
+	}
+	pod := n.lastPod
+	n.lastPod = nil
+
+	names := strings.Split(pod.Annotations["ci-operator.openshift.io/container-sub-tests"], ",")
+	if len(names) == 1 && names[0] == "" {
+		return nil
+	}
+	var tests []*junit.TestCase
+	for _, status := range pod.Status.ContainerStatuses {
```

Comment:
No check for `""` in the comma-delimited list?

Reply:
I can, it couldn't cause a match (since we wouldn't find a container with "")